### PR TITLE
Enable heterogeneous insert for static_map

### PR DESCRIPTION
**Pull request by sleeepyjack**
Analogous to #375 , this PR adds heterogeneous insert to `static_map`.

### CODE REVIEW

File: include/cuco/detail/static_map/static_map_ref.inl [hunk 1]
```diff
@@ -411,14 +427,17 @@ class operator_impl<
...
    * @param group The Cooperative Group used to perform group insert_and_find
    * @param value The element to insert
    *
    * @return a pair consisting of an iterator to the element and a bool indicating whether the
    * insertion is successful or not.
    */
+  template <typename Value>
   __device__ thrust::pair<iterator, bool> insert_and_find(
-    cooperative_groups::thread_block_tile<cg_size> const& group, value_type const& value) noexcept
+    cooperative_groups::thread_block_tile<cg_size> const& group, Value const& value) noexcept
```
Comment (PointKernel) [inline include/cuco/detail/static_map/static_map_ref.inl:480]: The non-CG `insert_and_find` needs to be updated as well.

**Review by PointKernel — CHANGES_REQUESTED** [review]
Just one leftover for scalar `insert_and_find` otherwise LGTM

File: include/cuco/detail/static_map/static_map_ref.inl (hunk 1 above)
Comment (sleeepyjack) [reply include/cuco/detail/static_map/static_map_ref.inl:480]: Addressed in [63e1fae](https://github.com/NVIDIA/cuCollections/pull/381/commits/63e1fae7af04eafeaa737ba175bfd87d0f51fc1f)

File: tests/static_map/heterogeneous_lookup_test.cu [hunk 2]
```diff
@@ -64,61 +66,74 @@ struct key_triplet {
...
+  auto my_map                    = cuco::experimental::static_map<Key,
+                                               Value,
+                                               cuco::experimental::extent<std::size_t>,
+                                               cuda::thread_scope_device,
+                                               custom_key_equal,
+                                               probe_type>{capacity,
+                                                           cuco::empty_key<Key>{sentinel_key},
+                                                           cuco::empty_value{sentinel_value},
+                                                           custom_key_equal{},
+                                                           probe};
```
Comment (PointKernel) [inline tests/static_map/heterogeneous_lookup_test.cu]:
```suggestion
  auto my_map                    = cuco::experimental::static_map{capacity,
                                                           cuco::empty_key<Key>{sentinel_key},
                                                           cuco::empty_value<Value>{sentinel_value},
                                                           custom_key_equal{},
                                                           probe};
```
nit: CTAD would work just fine.

**Review by PointKernel — APPROVED** [review]
LGTM

File: tests/static_map/heterogeneous_lookup_test.cu (hunk 2 above)
Comment (sleeepyjack) [reply tests/static_map/heterogeneous_lookup_test.cu]: Beau-ti-ful! Addressed in [cd03620](https://github.com/NVIDIA/cuCollections/pull/381/commits/cd03620c28be7a71bd798081ac9d650166177ae2)

File: include/cuco/detail/pair/pair.inl [hunk 3]
```diff
@@ -49,3 +49,11 @@ __host__ __device__ constexpr bool operator==(cuco::pair<T1, T2> const& lhs,
...
 
 }  // namespace cuco
+
+namespace thrust {
+#include <cuco/detail/pair/tuple_helpers.inl>
+}  // namespace thrust
+
+namespace cuda::std {
+#include <cuco/detail/pair/tuple_helpers.inl>
+}  // namespace cuda::std
```
Comment (jrhemstad) [inline include/cuco/detail/pair/pair.inl:59]: I think the namespaces should be in the headers

Comment (jrhemstad) [reply include/cuco/detail/pair/pair.inl:59]: Oh, I see, this is your trick to reuse the same specializations for Thrust and libcu++.

**Review by jrhemstad — APPROVED** [review]

**Pull request merged**